FWCore/Common/src/OutputProcessBlockHelper.cc

```diff
@@ -0,0 +1,231 @@
+        if (iStoredProcess < nInputProcesses) {
+          unsigned int cacheIndex = innerVectorOfCacheIndices[translateFromStoredIndex_[iStoredProcess]];
+          if (cacheIndex != invalidCacheIndex()) {
+            // The offsets count in the cache sequence to the first entry in
+            // the current input file and process
+            unsigned int inputOffset = fileOffset + processOffset[iStoredProcess];
+            unsigned int storedOffset = storedProcessOffset[iStoredProcess] + storedFileInProcessOffset[iStoredProcess];
+            storedCacheIndex = cacheIndex - inputOffset + storedOffset;
+          }
+        } else {
+          // This corresponds to the current process if it has newly produced
+          // ProcessBlock products (plus possibly SubProcesses).
+          storedCacheIndex = storedProcessOffset[nInputProcesses] + iStoredProcess - nInputProcesses;
+        }
+        storedCacheIndices.push_back(storedCacheIndex);
+      }
+    }
+  }
+
+  void OutputProcessBlockHelper::setStoredProcessOffset(unsigned int nInputProcesses,
+                                                        std::vector<std::vector<unsigned int>> const& nEntries,
+                                                        std::vector<unsigned int>& storedProcessOffset) const {
+    for (unsigned int iStored = 0; iStored < nInputProcesses + 1; ++iStored) {
+      storedProcessOffset[iStored] = 0;
+      // loop over earlier processes
+      for (unsigned int jStored = 0; jStored < iStored; ++jStored) {
+        unsigned int indexInEventProcessor = translateFromStoredIndex_[jStored];
+        // loop over input files
+        for (auto const& entries : nEntries) {
+          assert(indexInEventProcessor < entries.size());
+          storedProcessOffset[iStored] += entries[indexInEventProcessor];
+        }
+      }
+    }
+  }
+
+  void OutputProcessBlockHelper::setProcessOffset(unsigned int iFile,
+                                                  unsigned int nInputProcesses,
+                                                  std::vector<std::vector<unsigned int>> const& nEntries,
+                                                  std::vector<unsigned int>& processOffset) const {
+    for (unsigned int iStored = 0; iStored < nInputProcesses; ++iStored) {
+      processOffset[iStored] = 0;
+      unsigned int iProcess = translateFromStoredIndex_[iStored];
+      for (unsigned int jProcess = 0; jProcess < iProcess; ++jProcess) {
+        processOffset[iStored] += nEntries[iFile][jProcess];
+      }
+    }
+  }
+
+  void OutputProcessBlockHelper::setStoredFileInProcessOffset(
+      unsigned int iFile,
+      unsigned int nInputProcesses,
+      std::vector<std::vector<unsigned int>> const& nEntries,
+      std::vector<unsigned int>& storedFileInProcessOffset) const {
+    for (unsigned int iStored = 0; iStored < nInputProcesses; ++iStored) {
+      storedFileInProcessOffset[iStored] = 0;
```

Comment:
same here

Reply:
I changed the loop to a ranged for loop. Same comments as above.